chat/attachment.py

```diff
@@ -0,0 +1,275 @@
+            "max_image_height",
+            "max_image_file_size",
+            "default_image_quality",
+            "strict_image_quality",
+            "low_quality_threshold",
+            "low_quality_token_threshold",
+            "allowed_image_models",
+            "allowed_image_extensions",
+            "discord_url_allowed",
+        ),
+    )
+
+    def __init__(
+        self,
+        attachment: Attachment,
+        *,
+        quality: str = DEFAULT_QUALITY,
+        strict: bool = False,
+    ):
+        super().__init__(attachment)
+
+        self._strict = strict
+        if not strict and self.STRICT_QUALITY is not None:
+            self._quality = self.STRICT_QUALITY
+        else:
+            self._quality = quality
+            if not strict:
+                width, height = attachment.width, attachment.height
+                cost = self.calc_openai_tokens(width, height, quality=quality)
+
+                if max(width, height) > self.LOW_QUALITY_THRESHOLD:
+                    self._quality = "low"
+                if cost > self.LOW_QUALITY_TOKEN_THRESHOLD:
+                    self._quality = "low"
+
+        self._tokens = self.calc_openai_tokens(
+            attachment.width, attachment.height, quality=self._quality
+        )
+
+    @staticmethod
+    def calc_openai_tokens(width, height, *, quality: str):
+        from options import OPENAI_DEFAULT_IMAGE_COST, OPENAI_IMAGE_BLOCK_COST
+
+        default_cost, block_cost = OPENAI_DEFAULT_IMAGE_COST, OPENAI_IMAGE_BLOCK_COST
+
+        if width == 0 and height == 0:
+            return 0
+        elif quality == "low":
+            return default_cost
+        elif quality == "high":
+            cost = default_cost
+            if max(width, height) <= 512:
+                return cost + block_cost
+
+            ratio, ratio_inv = width / height, height / width
+
+            if max(width, height) > 2048:
+                if ratio > 1:
+                    width = 2048
+                    height = math.floor(2048 * ratio_inv)
+                elif ratio < 1:
+                    width = math.floor(2048 * ratio)
+                    height = 2048
+                else:
+                    width = height = 2048
+
+            if min(width, height) <= 768:
+                cost += block_cost * math.ceil(width / 512) * math.ceil(width / 512)
+            elif ratio > 1:
+                cost += block_cost * 2 * math.ceil(math.floor(768 * ratio) / 512)
+            elif ratio < 1:
+                cost += block_cost * 2 * math.ceil(math.floor(768 * ratio_inv) / 512)
+            else:
+                cost += block_cost * 4
+
+            return cost
+        elif quality == "auto":
+            if width <= 512 and height <= 512:
+                return default_cost
+            else:
+                return GPTImageAttachment.calc_openai_tokens(
+                    width, height, quality="high"
+                )
+        else:
+            return math.inf
```

Comment:
```suggestion
            raise ValueError(f"invalid quality: only low, high, and auto allowed")
```

enum을 선언해서 넘기는게 더 좋겠네요.